### PR TITLE
Update clrmd to 3.1

### LINKED ISSUE #2383: `DisassemblyDiagnoser` does not work in `InProcessEmitToolchain`
Trying to use `DisassemblyDiagnoser` with `InProcessEmitToolchain`, I get an error:

```cmd
// ! InProcessEmitExecutor, exception: System.InvalidOperationException: Sequence contains no matching element
   at System.Linq.ThrowHelper.ThrowNoMatchException()
   at System.Linq.Enumerable.First[TSource](IEnumerable`1 source, Func`2 predicate)
   at BenchmarkDotNet.Disassemblers.ClrMdV2Disassembler.AttachAndDisassemble(Settings settings)
   at BenchmarkDotNet.Disassemblers.SameArchitectureDisassembler.Disassemble(DiagnoserActionParameters parameters)
   at BenchmarkDotNet.Diagnosers.DisassemblyDiagnoser.Handle(HostSignal signal, DiagnoserActionParameters parameters)
   at BenchmarkDotNet.Diagnosers.CompositeDiagnoser.Handle(HostSignal signal, DiagnoserActionParameters parameters)
   at BenchmarkDotNet.Toolchains.InProcess.InProcessHost.SendSignal(HostSignal hostSignal)
   at BenchmarkDotNet.Engines.HostExtensions.AfterAll(IHost host)
   at BenchmarkDotNet.Toolchains.InProcess.Emit.Implementation.RunnableProgram.Run(BenchmarkId benchmarkId, Assembly partitionAssembly, BenchmarkCase benchmarkCase, IHost host)
   at BenchmarkDotNet.Toolchains.InProcess.Emit.InProcessEmitExecutor.ExecuteCore(IHost host, ExecuteParameters parameters)
Executable C:\Users\Tim\Documents\git\ProtoPromiseBenchmarks\AsynchronousBenchmarks\bin\Release\net6.0\001a3848-5fde-4a5d-a284-00d6519a2aaeEmitted.dll not found
ExitCode != 0 and no results reported
No Workload Results were obtained from the run.
```

The last version where it worked was 0.13.1. It seems `ClrMdV2Disassembler` broke this case in 0.13.2. @adamsitnik

### PR DESCRIPTION
**Pull request by timcassell**
Fixes #2383 

I left the x86/x64 disassemblers on v1, those can be updated separately.

Also, as discussed in #2383, this breaks netcoreapp3.0 and older, so we should probably release this in 0.14.0.

### CODE REVIEW

File: src/BenchmarkDotNet/Disassemblers/ClrMdV3Disassembler.cs [hunk 1]
```diff
@@ -213,60 +215,30 @@ private IEnumerable<Asm> Decode(ILToNativeMap map, State state, int depth, ClrMe
...
-                // Since we care about Tier 1 (if it's present), we "fake" a Tier 1 map.
-                return new[] { new ILToNativeMap() { StartAddress = startAddress, EndAddress = endAddress } };
-            }
-            else if (sortedMaps[0].StartAddress != startAddress || (sortedMaps[sortedMaps.Length - 1].EndAddress != endAddress && endAddress != ulong.MaxValue))
-            {
-                // In such situation ILOffsetMap most likely is missing few bytes. We just "extend" it to avoid producing "bad" instructions.
-                return new[] { new ILToNativeMap() { StartAddress = startAddress, EndAddress = endAddress } };
-            }
-
-            return sortedMaps;
```
Comment (adamsitnik) [inline src/BenchmarkDotNet/Disassemblers/ClrMdV3Disassembler.cs:247]: It took me a lot of time to workaround all ClrMd bugs. If we really want to remove these workarounds we need to do a very extensive testing: 

1. Run a simple benchmark and ask BDN to disassemble ALL methods for main. It should produce a markdown file with all CLR methods disassembled.
2. Same as above, but for this PR.
3. Diff two produced .md files and ensure that all instructions are correctly resolved and there are no missing things.

**Review by adamsitnik — COMMENTED** [review]
I am supportive of this change, but we need to perform a LOT of testing.

Thank you for your hard work on this @timcassell !

**Comment by timcassell** [comment]
@adamsitnik I ran `dotnet run -c Release -f net8.0 --filter BenchmarkDotNet.Samples.IntroDisassembly.SumLocal --disasmFilter '*'`
And it looks like we get a lot more information in 3.1 (as expected since clrmd was specifically updated to account for changes in the runtime).

image

I then ran the same benchmark with `net6.0` to compare the older runtime results. The output was in a different order, but after comparing individual functions, they look identical.

image

**Review by AndreyAkinshin — APPROVED** [review]
I have tested this branch with *some* benchmarks, and it works like a charm! No concerns from my side: I haven't found any issues. However, I'm not aware of all the corner cases and I don't know how to perform proper extensive testing. I approve the PR and I don't mind including it in v0.14.0.

@adamsitnik what do you think? If you feel confident, we can merge it. If you have any questions, please provide a set of scenarios that should be tested.

P.S. I feel like it's just impossible to verify such changes properly. There is always a chance of getting a corner case with a regression. We can always just give it a try: release the new version, collect feedback, and fix any further issues once they are discovered.

**Comment by adamsitnik** [comment]
I won't have the time to test all possible scenarios in the near future. Since it fixes https://github.com/dotnet/BenchmarkDotNet/issues/2638#issuecomment-2350873403 lets merge it now and wait for new potential bug reports.

**Pull request merged**